File: libs/java/common/src/main/java/arda/common/context/ArdaContext.java
package arda.common.context;

import java.util.Collections;
import java.util.List;
import java.util.Optional;

public record ArdaContext(
    String userId,
    String traceId,
    String tenantId,
    List<String> roles
) {
    // ThreadLocal for Imperative/Virtual Threads flow
    private static final ThreadLocal<ArdaContext> THREAD_LOCAL = new ThreadLocal<>();

    public ArdaContext {
        if (roles == null) roles = Collections.emptyList();
    }

    public static ArdaContext empty() {
        return new ArdaContext(null, null, null, Collections.emptyList());
    }

    /**
     * For Imperative/Virtual Threads flow (Standard for Java 25+).
     */
    public static ArdaContext current() {
        return Optional.ofNullable(THREAD_LOCAL.get()).orElse(empty());
    }

    public static void set(ArdaContext context) {
        THREAD_LOCAL.set(context);
    }

    public static void clear() {
        THREAD_LOCAL.remove();
    }

    public static String getTraceId() {
        return Optional.ofNullable(current().traceId()).orElse("unknown");
    }
}

File: arda-be-java/libs/common/src/main/kotlin/arda/common/context/ArdaContext.kt
package arda.common.context

import reactor.core.publisher.Mono
import reactor.util.context.Context

/**
 * Standardized Request Context for Arda Platform.
 * Follows Cloud-native patterns for context propagation in Reactive systems.
 */
data class ArdaContext(
    val userId: String? = null,
    val traceId: String? = null,
    val tenantId: String? = null,
    val roles: List<String> = emptyList()
) {
    companion object {
        private val CONTEXT_KEY = ArdaContext::class.java

        /**
         * Get the current ArdaContext from Reactor Context.
         */
        fun current(): Mono<ArdaContext> {
            return Mono.deferContextual { ctx ->
                if (ctx.hasKey(CONTEXT_KEY)) {
                    Mono.just(ctx.get(CONTEXT_KEY))
                } else {
                    Mono.just(ArdaContext())
                }
            }
        }

        /**
         * Create a Reactor Context with ArdaContext.
         */
        fun withContext(context: ArdaContext): Context {
            return Context.of(CONTEXT_KEY, context)
        }

        /**
         * Helper to get traceId for logging.
         */
        fun getTraceId(): Mono<String?> = current().map { it.traceId }
    }
}
